service/src/main/java/org/mskcc/smile/service/impl/TempoServiceImpl.java

```diff
@@ -152,4 +151,60 @@ private Tempo getDetailedTempoData(Tempo tempo) throws Exception {
     public void updateSampleBilling(SampleBillingJson billing) throws Exception {
         tempoRepository.updateSampleBilling(billing);
     }
+
+    private LocalDate getInitialPipelineRunDateBySamplePrimaryId(String primaryId) throws Exception {
+        String dateString = tempoRepository.findInitialPipelineRunDateBySamplePrimaryId(primaryId);
+        if (StringUtils.isEmpty(dateString)) {
+            LOG.warn("No Initial Pipeline Run Date found for sample with Primary ID: " + primaryId);
+            return null;
+        }
+        DateTimeFormatter runDateFormat = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+        LocalDateTime initialPipelineRunDate = LocalDateTime.parse(dateString, runDateFormat);
+        return initialPipelineRunDate.toLocalDate(); // return date only
+    }
+
+    private void populateTempoData(SmileSample sample, Tempo tempo) throws Exception {
+        SmileRequest request = requestService.getRequestBySample(sample);
+        String custodianInformation = Strings.isBlank(request.getLabHeadName())
+                ? request.getLabHeadName() : request.getInvestigatorName();
+        tempo.setCustodianInformation(custodianInformation);
+
+        String accessLevel = tempo.getAccessLevel();
+        String primaryId = sample.getPrimarySampleAlias();
+        LocalDate initialPipelineRunDate = getInitialPipelineRunDateBySamplePrimaryId(primaryId);
+
+        if (initialPipelineRunDate != null) {
+            LocalDate embargoDate = initialPipelineRunDate.plusMonths(EMBARGO_PERIOD_MONTHS);
+            tempo.setInitialPipelineRunDate(initialPipelineRunDate.format(DATE_FORMATTER));
+            tempo.setEmbargoDate(embargoDate.format(DATE_FORMATTER));
+            if (!sampleIsMarkedAsPublic(accessLevel)) {
+                // we release the sample the day after the embargo ends (confirmed with PMs)
+                tempo.setAccessLevel(LocalDate.now().isAfter(embargoDate)
+                        ? ACCESS_LEVEL_PUBLIC : ACCESS_LEVEL_EMBARGO);
+            }
+        } else {
+            // explicitly set dates to empty strings if no initial pipeline run date is found
+            tempo.setInitialPipelineRunDate("");
+            tempo.setEmbargoDate("");
+            if (!sampleIsMarkedAsPublic(accessLevel)) {
+                tempo.setAccessLevel(ACCESS_LEVEL_EMBARGO);
+            }
```

Comment:
If the initial run date/embargo date is missing, I'm setting access level to "MSK Embargo"... Alternatively, we could set access level as an empty string and set up an alert to check with the TEMPO team. What do you think?

Reply:
this makes sense to me